src/pudl/transform/sec10k.py

```diff
@@ -468,8 +606,52 @@ def core_sec10k__quarterly_exhibit_21_company_ownership(
     ).assign(
         filename_sec10k=lambda x: _simplify_filename_sec10k(x["filename_sec10k"]),
         fraction_owned=lambda x: _compute_fraction_owned(x["ownership_percentage"]),
-        report_date=lambda x: _year_quarter_to_date(x["year_quarter"]),
+        subsidiary_company_location=lambda x: _standardize_location(
+            x["subsidiary_company_location"]
+        ),
+    )
+    df["subsidiary_company_name"] = _remove_bad_subsidiary_names(
+        df["subsidiary_company_name"].str.strip().str.lower()
+    )
+    # a record is not meaningful if the subsidiary name is null
+    df = df.dropna(subset=["subsidiary_company_name"])
+    # construct the CIK of the parent company in order to
+    # create subsidiary_company_id_sec10k
+    df.loc[:, "parent_company_central_index_key"] = (
+        df["filename_sec10k"].apply(lambda x: x.split("/")[0]).str.zfill(10)
+    )
+    # sometimes there are subsidiaries with the same name but different
+    # locations of incorporation listed in the same ex. 21, so include
+    # location in the ID
+    df.loc[:, "subsidiary_company_id_sec10k"] = (
+        df["parent_company_central_index_key"]
+        + "_"
+        + df["subsidiary_company_name"]
+        + "_"
+        + df["subsidiary_company_location"].fillna("")
+    )
+    df = df.drop(columns="parent_company_central_index_key")
+    # filename_sec10k and subsidiary_company_id_sec10k should be a primary key
+    # for this table, however there are many instances where the model erroneously
+    # created two different records for the same subsidiary and location in
+    # one Ex. 21 attachment. Often, the fraction_owned field is null for one
+    # record and not the other. Keep only one of these records for each
+    # subsidiary_company_id_sec10k and filename_sec10k
+    n_dupes = len(
+        df[df.duplicated(subset=["filename_sec10k", "subsidiary_company_id_sec10k"])]
     )
+    logger.info(
+        f"""There are {n_dupes} / {len(df)} ({n_dupes / len(df)}) """
+        f"""duplicate filename_sec10k and subsidiary_company_id_sec10k records """
+        f"""from the extracted Ex. 21 ownership data which are dropped."""
+    )
+    df = (
+        df.sort_values(by=["fraction_owned"])
+        .groupby(["filename_sec10k", "subsidiary_company_id_sec10k"])
+        .first()
+        .reset_index()
+    )
```

Comment:
I added this to enforce `filename_sec10k` and `subsidiary_company_id_sec10k` as the PK for this table (and subsequently the PK for `out_sec10k__parents_and_subsidiaries`. This drops .4% of records which are erroneously created by the model. Each subsidiary company name and location pair should be unique to a filename (see commented lines above).